### PR TITLE
Count the uses of property, and do not generate properties that are not used

### PR DESCRIPTION
**Pull request by ogoffart**

### CODE REVIEW

**Comment by ogoffart** [comment]
I also tried to inline the properties that were used only once.  But this only inlined a handful of properties, and they were some bugs. So it wasn't really worth it

File: internal/compiler/llr/optim_passes/count_property_use.rs [hunk 1]
```diff
@@ -0,0 +1,116 @@
+// Copyright © SixtyFPS GmbH <info@slint-ui.com>
+// SPDX-License-Identifier: GPL-3.0-only OR LicenseRef-Slint-commercial
+
+//! Passes that fills the Property::use_count
+//!
+//! This pass assume that use_count of all properties is
```
Comment (tronical) [inline internal/compiler/llr/optim_passes/count_property_use.rs:6]:
```suggestion
//! This pass assume that use_count of all properties is zero
```

(guessing here :-)

File: internal/compiler/llr/optim_passes/count_property_use.rs [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+
+use crate::llr::{EvaluationContext, Expression, ParentCtx, PropertyReference, PublicComponent};
+
+pub fn count_property_use(root: &PublicComponent) {
+    // Visit the root property that are used.
```
Comment (tronical) [inline internal/compiler/llr/optim_passes/count_property_use.rs:11]:
```suggestion
    // Visit the root properties that are used.
```

File: internal/compiler/llr/optim_passes/count_property_use.rs [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+
+fn visit_property(pr: &PropertyReference, ctx: &EvaluationContext) {
+    let p_info = super::inline_expressions::property_binding_and_analysis(ctx, pr);
+    if let Some(p) = &p_info.property_decl {
+        p.use_count.set(p.use_count.get() + 1);
+    }
+    if let Some((binding, map)) = &p_info.binding {
+        let c = binding.use_count.get();
+        binding.use_count.set(c + 1);
+        if c == 0 {
```
Comment (tronical) [inline internal/compiler/llr/optim_passes/count_property_use.rs:98]: Do I understand correctly that the purpose of the zero check here is to make sure that we increase the use-count of properties referenced in the binding expression of `pr` only once and not on sub-sequent visitations?

File: internal/compiler/llr/optim_passes/count_property_use.rs [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+            if c > 0 {
+                continue;
+            }
+            match pr {
+                PropertyReference::Local { sub_component_path, property_index } => {
+                    let mut sc = sc;
+                    for i in sub_component_path {
+                        sc = &sc.sub_components[*i].ty;
+                    }
+                    if sc.properties[*property_index].use_count.get() == 0 {
```
Comment (tronical) [inline internal/compiler/llr/optim_passes/count_property_use.rs:37]: Is it guaranteed that `use_count` is up-to-date at this point? (i.e. the final value) What if the property's use count is increased later due to another use-site?

**Review by tronical — APPROVED** [review]

File: internal/compiler/llr/optim_passes/count_property_use.rs (hunk 3 above)
Comment (ogoffart) [reply internal/compiler/llr/optim_passes/count_property_use.rs:98]: basicaly the same algorithm of a garbage collector: we start with the "roots" and visit all the properties and their bindings from there.   Once a binding has been visited, no need to revisit it.

File: internal/compiler/llr/optim_passes/count_property_use.rs (hunk 4 above)
Comment (ogoffart) [reply internal/compiler/llr/optim_passes/count_property_use.rs:37]: Not the final value, but good enough for what we care about here: is the value read by the sub component itself.
For local properties, we will go recursively to find them.

**Pull request merged**